### ISSUE TITLE
not-an-issue - just a thank you

### ISSUE THREAD

**Issue by FloWi** [Hammerspoon/hammerspoon#3163]
This is not an issue, just a way to say thank you, so please ~~burn~~ close after reading.

I was finally able to automate fixing my display setup. I used displayplacer and some scala - see this [issue](https://github.com/jakehilborn/displayplacer/issues/77#issuecomment-1066973505) for details.

I spent hours fiddling around with different hooks-, events- and launchd-approaches and wanted to give up and just use an automator task triggered by a keyboard shortcut. But then I found hammerspoon and got it running within an hour. 🎉  That's really great, since it bothered me almost every day.

So big thanks for your excellent tool - the eventhooks just work and the rest was very easy to debug since you also provide a console for logging.

here is my hammerspoon setup if anyone is interested - the scala code is a mess and I have stuff hardcoded for my setup.

```lua
function fixScreenSetupIfNecessary() 
  result, status, type, rc = hs.execute("~/programming/mac-automation/fix-display-setup/fix_displays", true)
  print(result)
end

hs.hotkey.bind({"cmd", "alt", "ctrl"}, "H", function()
  fixScreenSetupIfNecessary()
end)


function sleepWatch(eventType)
  if (eventType == hs.caffeinate.watcher.screensDidUnlock) then
    print("hs.caffeinate.watcher.screensDidUnlock")
    fixScreenSetupIfNecessary()
  end
end

local sleepWatcher = hs.caffeinate.watcher.new(sleepWatch)
sleepWatcher:start()
```

Cheers,
Florian

several failed attempts to automate included

- <https://stackoverflow.com/questions/6442364/running-script-upon-login-mac>
- sleepwatcher (didn't spend too much time with it)
- launchd (works only on login - not after unlock)
- loginHooks (don't seem to work in macOS 11)
- eventscripts

**Comment by cmsj** [comment]
Thanks! I'm curious what it is you're fixing about your displays?

**Comment by FloWi** [comment]
I have two external monitors of the same model and macOS quite often mixes their placement up. Seems like the id or the uuid is not stable and does not take the serial number into account. If you take a look into the issues of the displayplacer repo you'll see a lot of people having issues with that.

**Comment by cmsj** [comment]
@FloWi thanks for the extra info. Since we do have the ability to position displays within the "global coordinate space" I'm going to see if we can usefully add the ability to get serial numbers for displays, so in theory you wouldn't even need to shell out to displayplacer if you didn't want to. I've pushed a PR (#3164), but it's currently untested beyond my machine with my two displays.

**Comment by FloWi** [comment]
@cmsj Oh, how cool is that? Thanks!

If you take a look at the conversation in this issue https://github.com/jakehilborn/displayplacer/issues/77, you'll see, that "works on my machine" is an issue when it comes to serial numbers of external monitors. Let me know if you need someone to try something out.

**Comment by cmsj** [comment]
Yeah that's the one I was reading while thinking about this. We already expose UUIDs, but as people have discovered on that issue, they are not guaranteed to be stable. If I've done #3164 correctly, we should now also be exposing the serial number that IOKit is reading from the hardware. The one thing that's confusing me though is that the serial number for my main display doesn't match the one I see in System Information (and neither does the one for my second display, because System Information doesn't show a serial at all for that device!)

**Comment by FloWi** [comment]
Ah, let's see if IOKit is able to read the serial. Is it possible to install/run the binary from the ci-build of your PR?


Here is a snippet that people can use to print out some details about their displays.
```lua
hs.hotkey.bind({"cmd", "alt", "ctrl"}, "I", function()
  for key, screen in pairs(hs.screen.allScreens()) do
    print("===================")
    print("name:" .. screen:name())
    print("id: " .. screen:id())
    print("uuid: " .. screen:getUUID())
    -- print("serial: " .. screen:getSerial())
  end
end)
```

**Comment by cmsj** [comment]
Closing this issue, since we worked out some nice improvements in the associated PR.

**Comment by jpike88** [comment]
Unfortunately the PR didn't work for me. CGDisplayIOServicePort returns 0, so ```print(screen:getInfo())``` returns nil for all external screens.

I can't believe Apple has failed so miserably on what shouldn't be rocket science. I wish there was an Apple engineer we could reach out to, who will acknowledge the failure that is their display API, and fix the damn thing. I can't believe in 2022 I am still unable to match a consistent, static ID to a given monitor, and even with DisplayPort it's musical chairs sometimes. Damn it I'm so frustrated.

**Comment by jpike88** [comment]
```<NSCGSDisplay: 0x107414d60> id = 2 uuid = 539FDD1C-D593-4F80-A05D-6462660B7326 frame = {{0, 0}, {1920, 1080}} depth = 4 backingPixelsPerPoint = 1 devicePixelCounts = {1920, 1080} colorSpace = <CGColorSpace 0x1065265a0> (kCGColorSpaceICCBased; kCGColorSpaceModelRGB; Apple RGB) maxHDRValue = 1 maxPotentialHDRValue = 1 maxReferenceHDRValue = 0```

Aside from the uuid value, which can't be trusted to remain constant on sleep/wake, there's nothing of use.

When I run ```ioreg -lw0 | grep DisplayAttributes```, I get something that looks useful under the ```AlphanumericSerialNumber``` attribute, but I don't know any way to associate that with it's equivalent CGDisplay.

**Comment by FloWi** [comment]
Another thing that comes to mind, @jpike88. Does anyone know if booting a linux os from usb works on macs? I don't know how usable that is (lots of colleagues using linux also had some trouble with their display setup, graphics drivers etc). Maybe it recognizes the monitors out of the box and you can get some detailed debugging infos about them. Maybe the serial number is hidden in a magical place somewhere else, and this can also be somehow accessed within objc. 🤷‍♂️

But I completely agree. This is all stuff the OS should abstract away. I never really had to deal with binary protocols and that EDID stuff felt all very "dirty" and error prone ;-)

**Comment by FloWi** [comment]
I searched github for `AlphanumericSerialNumber` and found only one match :)
https://github.com/vuo/vuo/blob/main/library/VuoScreenCommon.mm#L142

Would someone with objective-c knowledge mind taking a look, if that code might be useful?

**Comment by latenitefilms** [comment]
This looks very helpful - it looks like Apple Silicon and Intel is treated differently which would explain a lot. I'll have a play tomorrow unless someone beats me to it.

**Comment by FloWi** [comment]
> This looks very helpful - it looks like Apple Silicon and Intel is treated differently which would explain a lot. 

Nice. It'd be so cool if we could fix this issue. 

Maybe they only tested it with the Apple monitor and assumed no one would be able to afford more than one of them :p

**Comment by jpike88** [comment]
I've forked and made a branch using the above code @FloWi

https://github.com/jpike88/hammerspoon/tree/display-bug-fix

I'll let you know if I can get it to do what I want. Shouldn't take long

**Comment by jpike88** [comment]
Sadly the above code is flawed, as it's possible (especially if the monitors are the same model) for this code to raise a false positive:

```ioregSerialNumber == displaySerialNumber
                        && ioregVendorNumber == displayVendorNumber
                        && ioregModelNumber == displayModelNumber```

My 4 Samsung monitors all have identical values for those variables, meaning that it's just fetching the first one off the list. Which brings us back to the same conclusion I had, which is that there isn't a clear unique link between NSScreenNumber and a given listing in the io registry. Apple really screwed this up.

**Comment by jpike88** [comment]
```    | |   |   |   "DisplayAttributes" = {"SupportsSuspend"=No,"MaximumRefreshRate"=75,"SupportsActiveOff"=No,"PortID"=16,"ProductAttributes"={"ManufacturerID"="SAM","YearOfManufacture"=2019,"SerialNumber"=1515867474,"ProductName"="S24F350","AlphanumericSerialNumber"="H4ZM501352","LegacyManufacturerID"=19501,"ProductID"=3360,"WeekOfManufacture"=19},"MaxVerticalImageSize"=29,"MaxHorizontalImageSize"=52,"HasHDMILegacyEDID"=Yes,"Chromaticity"={"Red"={"X"=42624,"Y"=21824},"Green"={"X"=20736,"Y"=40128},"Blue"={"X"=9984,"Y"=4416}},"DefaultColorSpaceIsSRGB"=No,"NativeFormatHorizontalPixels"=1920,"DefaultWhitePoint"={"X"=20544,"Y"=21568,"Gamma"=144179},"SupportsVariableRefreshRate"=No,"AspectRatio"=15,"MinimumRefreshRate"=50,"WhitePoints"=({"X"=20544,"Y"=21568,"Gamma"=144179}),"PreciseAspectRatio"=117513,"ContinuousFrequencySupport"="None","SupportsStandby"=Yes,"NativeFormatVerticalPixels"=1080}
    | |   |   |   "DisplayAttributes" = {"SupportsSuspend"=No,"MaximumRefreshRate"=75,"SupportsActiveOff"=No,"PortID"=0,"ProductAttributes"={"ManufacturerID"="SAM","YearOfManufacture"=2016,"SerialNumber"=1515867474,"ProductName"="S24F350","AlphanumericSerialNumber"="H4ZH900191","LegacyManufacturerID"=19501,"ProductID"=3360,"WeekOfManufacture"=38},"MaxVerticalImageSize"=29,"MaxHorizontalImageSize"=52,"HasHDMILegacyEDID"=Yes,"Chromaticity"={"Red"={"X"=42624,"Y"=21824},"Green"={"X"=20736,"Y"=40128},"Blue"={"X"=9984,"Y"=4416}},"DefaultColorSpaceIsSRGB"=No,"NativeFormatHorizontalPixels"=1920,"DefaultWhitePoint"={"X"=20544,"Y"=21568,"Gamma"=144179},"SupportsVariableRefreshRate"=No,"AspectRatio"=15,"MinimumRefreshRate"=50,"WhitePoints"=({"X"=20544,"Y"=21568,"Gamma"=144179}),"PreciseAspectRatio"=117513,"ContinuousFrequencySupport"="None","SupportsStandby"=Yes,"NativeFormatVerticalPixels"=1080}
    | |   |   |   "DisplayAttributes" = {"SupportsSuspend"=No,"MaximumRefreshRate"=75,"SupportsActiveOff"=No,"PortID"=32,"ProductAttributes"={"ManufacturerID"="SAM","YearOfManufacture"=2016,"SerialNumber"=1515867474,"ProductName"="S24F350","AlphanumericSerialNumber"="H4ZH801712","LegacyManufacturerID"=19501,"ProductID"=3360,"WeekOfManufacture"=35},"MaxVerticalImageSize"=29,"MaxHorizontalImageSize"=52,"HasHDMILegacyEDID"=Yes,"Chromaticity"={"Red"={"X"=42624,"Y"=21824},"Green"={"X"=20736,"Y"=40128},"Blue"={"X"=9984,"Y"=4416}},"DefaultColorSpaceIsSRGB"=No,"NativeFormatHorizontalPixels"=1920,"DefaultWhitePoint"={"X"=20544,"Y"=21568,"Gamma"=144179},"SupportsVariableRefreshRate"=No,"AspectRatio"=15,"MinimumRefreshRate"=50,"WhitePoints"=({"X"=20544,"Y"=21568,"Gamma"=144179}),"PreciseAspectRatio"=117513,"ContinuousFrequencySupport"="None","SupportsStandby"=Yes,"NativeFormatVerticalPixels"=1080}
    | |   |   |   "DisplayAttributes" = {"SupportsSuspend"=No,"MaximumRefreshRate"=75,"SupportsActiveOff"=No,"PortID"=48,"ProductAttributes"={"ManufacturerID"="SAM","YearOfManufacture"=2019,"SerialNumber"=1515867474,"ProductName"="S24F350","AlphanumericSerialNumber"="H4ZM501355","LegacyManufacturerID"=19501,"ProductID"=3360,"WeekOfManufacture"=19},"MaxVerticalImageSize"=29,"MaxHorizontalImageSize"=52,"HasHDMILegacyEDID"=Yes,"Chromaticity"={"Red"={"X"=42624,"Y"=21824},"Green"={"X"=20736,"Y"=40128},"Blue"={"X"=9984,"Y"=4416}},"DefaultColorSpaceIsSRGB"=No,"NativeFormatHorizontalPixels"=1920,"DefaultWhitePoint"={"X"=20544,"Y"=21568,"Gamma"=144179},"SupportsVariableRefreshRate"=No,"AspectRatio"=15,"MinimumRefreshRate"=50,"WhitePoints"=({"X"=20544,"Y"=21568,"Gamma"=144179}),"PreciseAspectRatio"=117513,"ContinuousFrequencySupport"="None","SupportsStandby"=Yes,"NativeFormatVerticalPixels"=1080}```

Just to elaborate: The above 4 listings are dumped when I run ```ioreg -lw0 | grep DisplayAttributes```.  As you can see, each listing is indistinguishable from the next, except for that handy ```AlphanumericSerialNumber``` property. The problem, which the above code link tries to do, is search the listings and try to match the ```SerialNumber```, ```LegacyManufacturerID```, ```ProductID``` properties with their ```CGDisplay``` equivalents. Not a very useful way if you have multiple monitors that are the same for those values.

**Comment by FloWi** [comment]
I would ignore `displaySerialNumber` part for now and see if this works on your machine.
```
serialString = [[properties[@"DisplayAttributes"][@"ProductAttributes"][@"AlphanumericSerialNumber"] retain] autorelease];
```

I noticed that the manufactoring date is different for all your monitors. (`YearOfManufacture` and `WeekOfManufacture`)

Could you just use that to map to the correct monitors? @cmsj noticed that for my setup, so maybe it helps you as well :)

**Comment by FloWi** [comment]
@jpike88 I _might_ have found a link between the `DisplayID` and the `IODisplayAttributes`. It's more like a trail of breadcrumbs, but hey 🤷‍♂️ 

During the [discussion](https://github.com/Hammerspoon/hammerspoon/pull/3164#issuecomment-1073299881) @cmsj pointed out, that I might use `IODisplayLocation`. That string looked ugly so I discarded that ;-)

During my investigation I found this stackoverflow [answer](https://stackoverflow.com/questions/50238405/objective-c-get-displayproductid-and-displayvendorid-in-macos-programmatically) referring to this: `com.apple.windowserver.plist`

Here is my example (brace for impact though):
My main (external) monitor has the displayID `731392266`

```bash
defaults read /Library/Preferences/com.apple.windowserver.plist | egrep "IODisplayLocation|DisplayID"
```

Here you get the match between the displayID and the IODisplayLocation (which you also might get in the objc-world.

```text
DisplayID = 731392266;
IODisplayLocation = "IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/PEG0@1/IOPP/GFX0@0/ATY,Palena@4/AMDFramebufferVIB";
```

using that location `IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/PEG0@1/IOPP/GFX0@0/ATY,Palena@4/AMDFramebufferVIB` you can get a unique match in `ioreg` in the node `IODisplayPrefsKeyOld`. 

```bash
ioreg -lw0 | grep "IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/PEG0@1/IOPP/GFX0@0/ATY,Palena@4/AMDFramebufferVIB" -A20 -B20 | grep IODisplayAttributes
```

```text
  +-o AppleDisplay  <class AppleDisplay, id 0x1000044f2, registered, matched, active, busy 0 (0 ms), retain 8>
{
  "DisplayProductID" = 24740
  "IODisplayEDID" = <00ffffffffffff0030aea4600101010129170104b54028783e1df5ae4f35b3250d5054a10800810081809500a9c0b300d1c081c0d100e26800a0a0402e603020360081912100001a000000ff0056313138373032300a20202020000000fd0032471e641e010a202020202020000000fc004c454e204c543330353370774101df020321f14d90010203040607051112131f14230907078301000066030c001000107f2180e020403a6030403a0081912100001aad4300d051403a6060883a0081912100001c662150b051001b304070360081912100001edb6580c072403a6090d03a0081912100001c00000000000000000000000000000000000000000000d8>
  "IODisplayPrefsKeyOld" = "IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/PEG0@1/IOPP/GFX0@0/ATY,Palena@4/AMDFramebufferVIB/display0/AppleDisplay-30ae-60a4"
  "IODisplayPrefsKey" = "IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/PEG0@1/IOPP/GFX0@0-30ae-60a4-55bd00373315cc32"
  "DisplayVendorID" = 12462
  "IODisplayAttributes" = {"IODisplayAttributes"=<676174760000000073676c6600000000726c6f63110100006364670080808000676c666401000000>,"IODisplayVSDB030C00"=<100010>}
  "DisplaySerialNumber" = 0
}
```

On my system that info is not very useful, but you wrote that you were able to find your serial number in there.

And _of course_ there is the very useful `"DisplaySerialNumber" = 0` ;-)

**Comment by asmagill** [comment]
Don't know if this helps or not, but since `ioreg` has been mentioned I thought I'd note that it's among the stuff Apple open sourced -- https://github.com/apple-oss-distributions/IOKitTools/blob/main/ioreg.tproj/ioreg.c

Haven't had a chance to dig into it myself, so I couldn't say if it's doing something @cmsj's code isn't or not, but figured it might be worth a look.

**Comment by jpike88** [comment]
I don't have ```com.apple.windowserver.plist```, but rather ```com.apple.windowserver.displays.plist```

```
{
    DisplayAnyUserSets =     {
        Configs =         (
                        (
                                {
                    CurrentInfo =                     {
                        Depth = 8;
                        High = 1080;
                        Hz = 60;
                        IsLink = 0;
                        OriginX = 0;
                        OriginY = 0;
                        Rotation = 0;
                        Scale = 1;
                        Wide = 1920;
                    };
                    Rotation = 0;
                    UUID = "539FDD1C-D593-4F80-A05D-6462660B7326";
                    UnmirrorInfo =                     {
                        Depth = 8;
                        High = 1080;
                        Hz = 60;
                        IsLink = 0;
                        OriginX = 0;
                        OriginY = 0;
                        Rotation = 0;
                        Scale = 1;
                        Wide = 1920;
                    };
                },
                               

   ... another 6 more of the above item in here....

        );
        Orientations =         {
            "04E9A6E1-5FFA-4C3F-9D5D-F37093C8E811" = 0;
            "37D8832A-2D66-02CA-B9F7-8F30A301B230" = 0;
            "3985A09A-2E3F-45C6-AF49-4F0ACAEF938F" = 0;
            "462D4D76-FEEF-4004-80D4-1F7276D6762B" = 0;
            "539FDD1C-D593-4F80-A05D-6462660B7326" = 0;
            "8AE118C7-C8B6-42E8-A15D-039605B6A24D" = 0;
            "FFB292C7-79C7-4076-B01D-C257E4EAEEDF" = 0;
        };
        Underscan =         {
            "04E9A6E1-5FFA-4C3F-9D5D-F37093C8E811" = 1;
            "37D8832A-2D66-02CA-B9F7-8F30A301B230" = 1;
            "3985A09A-2E3F-45C6-AF49-4F0ACAEF938F" = 1;
            "462D4D76-FEEF-4004-80D4-1F7276D6762B" = 1;
            "539FDD1C-D593-4F80-A05D-6462660B7326" = 1;
            "8AE118C7-C8B6-42E8-A15D-039605B6A24D" = 1;
            "FFB292C7-79C7-4076-B01D-C257E4EAEEDF" = 1;
        };
        Version = 1;
    };
    DisplayUUIDMappings =     {
        "12C2775D-3A83-0F83-0F93-5A69417CA049" = "539FDD1C-D593-4F80-A05D-6462660B7326";
        "58CA9ABD-C523-0889-0653-0AD050454C5C" = "462D4D76-FEEF-4004-80D4-1F7276D6762B";
        "58CA9ABD-C523-0889-0F93-5A69417CA049" = "04E9A6E1-5FFA-4C3F-9D5D-F37093C8E811";
        "62CE3D68-6FFA-4F48-0F93-5A69417CA049" = "9A250AFF-31DA-491E-A123-E4B692D0B273";
        "80FBD496-C52B-7E94-0F93-5A69417CA049" = "3985A09A-2E3F-45C6-AF49-4F0ACAEF938F";
        "9B603E2E-98E1-E840-0F93-5A69417CA049" = "AE0FE03A-EFC0-45B1-9324-31D5604CF208";
        "F9633312-6BE0-7075-0F93-5A69417CA049" = "8AE118C7-C8B6-42E8-A15D-039605B6A24D";
        "FFD321D7-88E3-3DB6-0F93-5A69417CA049" = "FFB292C7-79C7-4076-B01D-C257E4EAEEDF";
    };
}
```

As we can see, it's relying entirely on these generated UUID values, which mean I can't get the values you're looking for. This must be an M1 Mac issue or something

**Comment by Thoro** [comment]
Maybe it's possible to match the ID of the object back?

```
    | |   | +-o AppleCLCD2  <class AppleCLCD2, id 0x10000031b, registered, matched, active, busy 0 (1162 ms), retain 11>
    | |   |   |   "DisplayAttributes" = {"SupportsSuspend"=No,"MaximumRefreshRate"=76,"SupportsActiveOff"=No,"SupportsBT2020cYCC"=64,"PortID"=1,"ProductAttributes"={"ManufacturerID"="BNQ","YearOfManufacture"=2020,"SerialNumber"=21573,"ProductName"="BenQ PD2720U","AlphanumericSerialNumber"="PAL02071019","LegacyManufacturerID"=2513,"ProductID"=32816,"WeekOfManufacture"=43},"MaxVerticalImageSize"=34,"SupportsHLGEOTF"=No,"MaxHorizontalImageSize"=60,"HasHDMILegacyEDID"=No,"Chromaticity"={"Red"={"X"=44032,"Y"=20864},"Green"={"X"=13568,"Y"=46720},"Blue"={"X"=9728,"Y"=3648}},"DefaultColorSpaceIsSRGB"=Yes,"SupportsBT2020YCC"=128,"NativeFormatHorizontalPixels"=3840,"DefaultWhitePoint"={"X"=20544,"Y"=21568,"Gamma"=144179},"SupportsHDRGammaEOTF"=No,"SupportsVariableRefreshRate"=No,"AspectRatio"=15,"MinimumRefreshRate"=50,"Luminance"={"Max"=23523611,"MaxFrameAverage"=23019548,"Min"=19809},"SupportsPQEOTF"=Yes,"WhitePoints"=({"X"=20544,"Y"=21568,"Gamma"=144179}),"PreciseAspectRatio"=115652,"ContinuousFrequencySupport"="None","SupportsStandby"=Yes,"SupportsBT2020RGB"=512,"SupportsHDRStaticMetadataType1"=Yes,"NativeFormatVerticalPixels"=2160}
    | |   |   |   "IOClass" = "AppleCLCD2"
    | |   | +-o AppleCLCD2  <class AppleCLCD2, id 0x100000352, registered, matched, active, busy 0 (1065 ms), retain 11>
    | |   |   |   "DisplayAttributes" = {"SupportsSuspend"=No,"MaximumRefreshRate"=76,"SupportsActiveOff"=No,"SupportsBT2020cYCC"=64,"PortID"=2,"ProductAttributes"={"ManufacturerID"="BNQ","YearOfManufacture"=2020,"SerialNumber"=21573,"ProductName"="BenQ PD2720U","AlphanumericSerialNumber"="PAL02081019","LegacyManufacturerID"=2513,"ProductID"=32816,"WeekOfManufacture"=43},"MaxVerticalImageSize"=34,"SupportsHLGEOTF"=No,"MaxHorizontalImageSize"=60,"HasHDMILegacyEDID"=No,"Chromaticity"={"Red"={"X"=44032,"Y"=20864},"Green"={"X"=13568,"Y"=46720},"Blue"={"X"=9728,"Y"=3648}},"DefaultColorSpaceIsSRGB"=Yes,"SupportsBT2020YCC"=128,"NativeFormatHorizontalPixels"=3840,"DefaultWhitePoint"={"X"=20544,"Y"=21568,"Gamma"=144179},"SupportsHDRGammaEOTF"=No,"SupportsVariableRefreshRate"=No,"AspectRatio"=15,"MinimumRefreshRate"=50,"Luminance"={"Max"=23523611,"MaxFrameAverage"=23019548,"Min"=19809},"SupportsPQEOTF"=Yes,"WhitePoints"=({"X"=20544,"Y"=21568,"Gamma"=144179}),"PreciseAspectRatio"=115652,"ContinuousFrequencySupport"="None","SupportsStandby"=Yes,"SupportsBT2020RGB"=512,"SupportsHDRStaticMetadataType1"=Yes,"NativeFormatVerticalPixels"=2160}
    | |   |   |   "IOClass" = "AppleCLCD2"
```

As you can see my two displays are identical, except for a single number in the serial :D

But there's these idd: 0x100000352, 0x10000031b. - And they stay consistent across ioreg calls... maybe it's possible to find it?

@jpike88 Exactly why those stupid monitors switch around, display manager just isn't aware that the monitor can change UUID - I think someone thought that those UUIDs are fixed, and never bothered to check - and it seems on M1 they change

**Comment by latenitefilms** [comment]
I'll try have a play on my Apple Silicon laptop and see if I can reproduce any of this.

Dumb idea... but have you given Hammerspoon Screen Recording permissions?

**Comment by FloWi** [comment]
I only have `com.apple.windowserver.plist`, and no `com.apple.windowserver.displays.plist`:-/

I googled a bit and found another location for plist files.

```bash
 ls -la ~/Library/Preferences/ByHost | grep window
```

```text
-rw-------  1 florian_witteler  staff   740B Mar 23 09:44 com.apple.loginwindow.82854E12-62D5-5A23-BADB-0243CB35190F.plist
-rw-r--r--  1 florian_witteler  admin    24K Jul 21  2013 com.apple.windowserver.0019e33b6e7d.plist
-rw-------  1 florian_witteler  admin    84K Nov  5  2018 com.apple.windowserver.3388F95F-88EF-50C8-A16E-9D03A516555A.plist
-rw-------  1 florian_witteler  admin    64K Apr 20  2016 com.apple.windowserver.56FFBA5A-DB19-5920-A45E-179DCE708CA4.plist
-rw-r--r--  1 florian_witteler  admin     0B Jul 21  2013 com.apple.windowserver.56FFBA5A-DB19-5920-A45E-179DCE708CA4.plist.lockfile
-rw-r--r--  1 florian_witteler  admin    24K Jul 21  2013 com.apple.windowserver.707554EB-2CB2-599D-ACE6-B78B05DCDC5D.plist
-rw-------@ 1 florian_witteler  staff    73K Mar 23 09:45 com.apple.windowserver.82854E12-62D5-5A23-BADB-0243CB35190F.plist
```

At least on my system (Intel mac with macOS 11.6.4). 

Most of the files are old, but if I take a look into the newest one, I can find a link between the `DisplayID` and `IODisplayLocation`

I wasn't able to read the file with `defaults read`, so I had to convert it to json with `plutil` (note - this tool _overwrites by default_ when you convert if you don't specify an output file 🤦‍♂️).

```bash
plutil -convert json /Users/florian_witteler/Library/Preferences/ByHost/com.apple.windowserver.82854E12-62D5-5A23-BADB-0243CB35190F.plist -o /tmp/plist.
json

cat /tmp/plist.json | jq ' [ .DisplaySets[][] | { DisplayID: .DisplayID, IODisplayLocation: .IODisplayLocation } ] | sort_by(.DisplayID) | unique '
```

```json
[
  {
    "DisplayID": 731392266,
    "IODisplayLocation": "IOService:/AppleACPIPlatformExpert/PCI0@0/AppleACPIPCI/PEG0@1/IOPP/GFX0@0/ATY,Palena@4/AMDFramebufferVIB"
  }
]
```

We could also have a pairing-session to stare at the mess together. I have no experience with obj-c but I'm interested in learning something new and maybe we can discover something.

**Comment by latenitefilms** [comment]
`CGDisplayIOServicePort` is no longer available on Apple Silicon and will always return 0.

`IODisplayConnect` is also no longer supported on Apple Silicon.

I did come across this comment:

> According to Apple Developer Support, `IODisplayConnect` is no longer supported on Apple M1 systems.
> 
> In order for me to get to the display information I had to look for `IODPDevice` to iterate on. Then get properties from the child service of the `IODPDevice`.
> 
> Be aware that the `IODPDevice` service does not contain the same data that `IODisplayConnect` did.

...but I couldn't get any useful information out of `IODPDevice` either.

I wonder if similar techniques could be used that were used for screen brightness on Apple Silicon Mac's (#2676)?

For future reference, here was my test code:

```objc
/// hs.screen:getASInfo() -> table or nil
/// Method
/// Gets a table of information about an `hs.screen` object on Apple Silicon.
///
/// Parameters:
///  * None
///
/// Returns:
///  *  A table containing various information, or nil if an error occurred.
///
/// Notes:
///  * Example usage: `hs.screen.primaryScreen():getASInfo()`
static int screen_getASInfo(lua_State *L) {
    LuaSkin *skin = [LuaSkin sharedWithState:L];
    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TBREAK];

    NSScreen *screen = get_screen_arg(L, 1);
    CGDirectDisplayID screen_id = [[[screen deviceDescription] objectForKey:@"NSScreenNumber"] intValue];
        
    [LuaSkin logError:[NSString stringWithFormat:@"hs.screen:getASInfo(): screen_id: %u", screen_id]];
    
    NSDictionary *deviceInfo = nil;
        
    io_iterator_t iter;
    io_service_t serv;
    io_service_t service = 0;

    //CFMutableDictionaryRef matching = IOServiceMatching("IODisplayConnect");
    
    CFMutableDictionaryRef matching = IOServiceMatching("IODPDevice");
    
    kern_return_t err = IOServiceGetMatchingServices(kIOMasterPortDefault, matching, &iter);
    if (err) {
        [LuaSkin logError:@"hs.screen:getASInfo(): matching failed."];
        return 0;
    }

    [LuaSkin logError:[NSString stringWithFormat:@"hs.screen:getASInfo(): iter: %u", iter]];
    
    while ((serv = IOIteratorNext(iter)) != 0) {
        [LuaSkin logError:@"hs.screen:getASInfo(): Checking screen..."];
        
        CFDictionaryRef info = IODisplayCreateInfoDictionary(serv, kIODisplayOnlyPreferredName);

        CFNumberRef vendorIDRef = CFDictionaryGetValue(info, CFSTR(kDisplayVendorID));
        CFNumberRef productIDRef = CFDictionaryGetValue(info, CFSTR(kDisplayProductID));

        CFIndex vendorID;
        CFIndex productID;
        Boolean success = CFNumberGetValue(vendorIDRef, kCFNumberCFIndexType, &vendorID);
        success &= CFNumberGetValue(productIDRef, kCFNumberCFIndexType, &productID);

        [LuaSkin logError:[NSString stringWithFormat:@"hs.screen:getASInfo(): vendorID: %ld", (long)vendorID]];
        [LuaSkin logError:[NSString stringWithFormat:@"hs.screen:getASInfo(): productID: %ld", (long)productID]];
        
        if (!success) {
            [LuaSkin logError:@"hs.screen:getASInfo(): not success"];
            CFRelease(info);
            continue;
        }

        if (CGDisplayVendorNumber(screen_id) != vendorID || CGDisplayModelNumber(screen_id) != productID) {
            [LuaSkin logError:@"hs.screen:getASInfo(): vendor and product don't match"];
            CFRelease(info);
            continue;
        }

        // Match!
        [LuaSkin logError:@"hs.screen:getASInfo(): match!!!"];
        service = serv;
        CFRelease(info);
        break;
    }

    IOObjectRelease(iter);
    
    if (service) {
        deviceInfo = (__bridge NSDictionary *)IODisplayCreateInfoDictionary(service, kIODisplayOnlyPreferredName);
    } else {
        [LuaSkin logError:@"hs.screen:getASInfo(): service was invalid"];
    }
    [skin pushNSObject:deviceInfo withOptions:LS_NSPreserveLuaStringExactly];
    return 1;
}
```

**Comment by latenitefilms** [comment]
Also worth noting...

> **SwitchResX 4.11.2**
> New for M1 Macs: Allows to read and export EDID on M1 Macs again (when possible)
> https://www.madrau.com/srx_download/srx_download/history.php
>

**Comment by jpike88** [comment]
Just tried SwitchResX, the same problem happens with it as well.

**Comment by jpike88** [comment]
Chatted to maker of SwitchResX, basically he's seeing the same limitation we're running into. All that's left is for Apple to fix this before the heat death of the universe.

A solution I'm going to play with in the meantime is a programmable EDID emulator/s. By spoofing different types of monitors, that should be enough for macOS to differentiate them. Maybe a little overkill but probably worth a go.

**Comment by FloWi** [comment]
> a programmable EDID emulator/s. By spoofing different types of monitors

Sounds interesting but I don't really understand what you're trying to do and how this might solve the issue. Would you mind to elaborate?

**Comment by jpike88** [comment]
It's too expensive to try, the idea is to fool macOS into thinking 4 totally different make/model monitors were connected to it, which may reduce the chances of it mixing them up

**Comment by FloWi** [comment]
Ah, I see. Sneaky ;-)
Too bad one doesn't really get support from apple. This all sounds like an easy fix for them to make :-(

**Comment by AvivWn** [comment]
I see that you work with ```sleepWatcher```. Is it keep working for every unlock? Cause I experiencing that it works for only few minute.

Which OS are you using and which hs version?

**Comment by hasantahir** [comment]
I would love to see any solution to this problem. Whenever I plug in my laptop to the external monitors in my office, I have to set up the messed display orientation of my identical monitors. 

Thoughts for novices @FloWi and @jpike88?

Cheers.

**Comment by jpike88** [comment]
Just plug in the right combo of power and data cables till it works.

I have this crazy theory, where I think the minuscule differences in how the monitors wake from standby help determine the 'stability' a multi-monitor positioning. However I'll add that two of my monitors have different year of manufacture to the others (which have the same), which in addition may have influence on how they're treated.

I've found the golden combination, and it seems to wake into the same lot each time, but not 100% of the time.

In other words I have no #$#%#% clue what to do, and someone in Apple deserves a hard slap.

**Comment by hasantahir** [comment]
Thanks. I will continue my pursuit of the magic combo :). Also, will gladly offer a helping hand in the last part.
